crates/cairo-profiler/src/main.rs

```diff
@@ -10,108 +10,80 @@ use crate::versioned_constants_reader::read_and_parse_versioned_constants_file;
-    let profiler_config = ProfilerConfig::from(&cli);
-
-    if profiler_config.show_inlined_functions
-        && !compiled_artifacts_cache.statements_functions_maps_are_present()
-    {
-        eprintln!(
-            "[\x1b[0;33mWARNING\x1b[0m] Mappings used for generating information about \
-        inlined functions are missing. Make sure to add this to your Scarb.toml:\n\
-        [profile.dev.cairo]\nunstable-add-statements-functions-debug-info = true"
-        );
-    }
-
-    let samples = collect_samples_from_trace(
-        &serialized_trace,
-        &compiled_artifacts_cache,
-        &profiler_config,
-        &os_resources_map,
-    )?;
-
-    let profile = build_profile(&samples);
-
-    if let Some(parent) = cli.output_path.parent() {
-        fs::create_dir_all(parent)
-            .context("Failed to create parent directories for the output file")?;
+            let VersionedCallTrace::V1(serialized_trace) =
+                serde_json::from_str(&data).context("Failed to deserialize call trace")?;
+
+            let compiled_artifacts_cache =
+                collect_and_compile_all_sierra_programs(&serialized_trace)?;
+            let profiler_config = ProfilerConfig::from(build_cli);
+
+            if profiler_config.show_inlined_functions
+                && !compiled_artifacts_cache.statements_functions_maps_are_present()
+            {
+                eprintln!(
+                    "[\x1b[0;33mWARNING\x1b[0m] Mappings used for generating information about \
+                inlined functions are missing. Make sure to add this to your Scarb.toml:\n\
+                [profile.dev.cairo]\nunstable-add-statements-functions-debug-info = true"
+                );
+            }
+
+            let samples = collect_samples_from_trace(
+                &serialized_trace,
+                &compiled_artifacts_cache,
+                &profiler_config,
+                &os_resources_map,
+            )?;
+
+            let profile = build_profile(&samples);
+
+            if let Some(parent) = build_cli.output_path.parent() {
+                fs::create_dir_all(parent)
+                    .context("Failed to create parent directories for the output file")?;
+            }
+            let mut file = fs::File::create(&build_cli.output_path)
+                .context("Failed to create the output file")?;
+
+            let mut buffer = BytesMut::new();
+            profile
+                .encode(&mut buffer)
+                .expect("Failed to encode the profile to the buffer");
+
+            let mut buffer_reader = buffer.reader();
+            let mut encoder = GzEncoder::new(&mut buffer_reader, Compression::default());
+
+            let mut encoded = vec![];
+            encoder
+                .read_to_end(&mut encoded)
+                .context("Failed to read bytes from the encoder")?;
+            file.write_all(&encoded).unwrap();
```

Comment:
[super nit]
```suggestion
            fs::write(&build_cli.output_path, &encoded)?;
```

This way we don't need:

```
          let mut file = fs::File::create(&build_cli.output_path)
                .context("Failed to create the output file")?;
```